### ISSUE TITLE
KunderaEntityTransaction throwing java.lang.IllegalStateException: Transaction is already active

### ISSUE THREAD

**Issue by rickcole3-yahoo** [Impetus/kundera#190]
We are seeing an issue when trying to run multiple calls to the entity manager from within a single transaction. We are using the integrated Spring transaction annotations with Kundera and Cassandra. We have a scenario where we do a query() and then a persist() within the same transaction. When we call entityManager.persist(), the subsequent calls to checkTransactionNeeded() and onLookUp() cause another call to KunderaEntityTransaction.begin(). Since the transaction is already under way, this causes the IllegalStateException to be thrown since the isActive() call returns true.

Also, I have noticed that there is a boolean flag in KunderaJTAUserTransaction for isTransactionInProgress that gets set to true once and is never gets set to false. We are directly binding an instnace the KunderJTAUserTransaction to JNDI. This is a global multiple thread accessed class. Should this boolean be used in this manner? Are we not binding the UserTransaction correctly?

**Comment by mevivs** [comment]
```
Also, I have noticed that there is a boolean flag in KunderaJTAUserTransaction for isTransactionInProgress that gets set to true once and is never gets set to false. We are directly binding an instnace the KunderJTAUserTransaction to JNDI. This is a global multiple thread accessed class. Should this boolean be used in this manner? Are we not binding the UserTransaction correctly?
```

I guess, this is where the issue is.  "isTransactionInProgress" is somehow not handled correctly in case of JTA. Should work fine for RESOURCE_LOCAL. 

I will add a fix and junit for this. Trivial one :)

-Vivek

**Comment by rickcole3-yahoo** [comment]
I guess my concern on this one is that the KunderaJTAUserTransaction can be accessed by multiple threads but the isTransactionInProgress flag is not synchronized or tied to any particular thread. This could cause some indeterminate results. Unless we are using it improperly? How is the KunderJTAUserTransaction generally bound to JNDI?

**Comment by mevivs** [comment]
Certainly, i agree and will modify it accordingly. Have a look at: 
https://github.com/impetus-opensource/Kundera/blob/trunk/kundera-core/src/main/java/com/impetus/kundera/persistence/jta/UserTransactionFactory.java

for more on binding. As long as multiple threads are joining in with in same JTA transaction boundary, "isTransactionInProgress" is bit error prone and need to be modified.

-Vivek

**Comment by mevivs** [comment]
Target for 2.4 release.

**Comment by mevivs** [comment]
A fix has been checked-in for this. Please verify.

-vivek

**Comment by rickcole3-yahoo** [comment]
This looks much happier now. Thanks!

**Comment by rickcole3-yahoo** [comment]
Well, after looking at the code, I think there is still a problem. The KunderaJTAUserTransaction instance is global which is accessed by multiple threads. The isTransactionInProgress flag can be modified by multiple threads so its status is indeterminate during runtime. It seems like the thread local transaction should be checked instead of the global KunderaJTAUserTransaction.isTransactionInProgress flag.

**Comment by mevivs** [comment]
I guess it makes much sense. Somehow i feel isTransactionInProgress doesn't add any value except troubling with  above mentioned scenarios. With recent changes i guess local transaction is already handling such scenarios.

Make sense?

-Vivek

**Comment by rickcole3-yahoo** [comment]
I agree that isTransactionInProgress flag should come out as it can cause some problems. Thanks for the quick responses, as usual.

**Comment by mevivs** [comment]
Fixed for synchronization as well. Junit added as well.

-Vivek

**Comment by mevivs** [comment]
Fixed and releasing with 2.5

-Vivek

**Comment by kavitacoded** [comment]
can anyone tell me this error? 
Exception in thread "main" java.lang.IllegalStateException: Transaction already active
	at org.hibernate.engine.transaction.internal.TransactionImpl.begin(TransactionImpl.java:74)
	at org.hibernate.internal.AbstractSharedSessionContract.beginTransaction(AbstractSharedSessionContract.java:471)
	at com.nt.test.SaveObjectTests.main(SaveObjectTests.java:26)
